src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotOverloadOperatorEqualsOnReferenceTypesTests.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Xunit;
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.DoNotOverloadOperatorEqualsOnReferenceTypes,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+using VerifyVB = Test.Utilities.VisualBasicCodeFixVerifier<
+    Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.DoNotOverloadOperatorEqualsOnReferenceTypes,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+
+namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.UnitTests
+{
+    public class DoNotOverloadOperatorEqualsOnReferenceTypesTests
+    {
+        [Fact]
+        public async Task OperatorEqual_ReferenceType_Diagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public class C
+{
+    public static bool operator ==(C left, C right) => true;
+    public static bool operator !=(C left, C right) => true;
+}",
+                VerifyCS.Diagnostic().WithSpan(4, 33, 4, 35).WithArguments("C"));
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+Public Class C
+    Public Shared Operator =(ByVal left As C, ByVal right As C) As Boolean
+        Return True
+    End Operator
+
+    Public Shared Operator <>(ByVal left As C, ByVal right As C) As Boolean
+        Return True
+    End Operator
+End Class",
+                VerifyVB.Diagnostic().WithSpan(3, 28, 3, 29).WithArguments("C"));
+        }
+
+        [Fact]
+        public async Task OperatorEqual_ValueType_NoDiagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public struct C
+{
+    public static bool operator== (C left, C right) => true;
+    public static bool operator !=(C left, C right) => true;
+}");
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+Public Structure C
+    Public Shared Operator =(ByVal left As C, ByVal right As C) As Boolean
+        Return True
+    End Operator
+
+    Public Shared Operator <>(ByVal left As C, ByVal right As C) As Boolean
+        Return True
+    End Operator
+End Structure");
+        }
+
+        [Fact]
+        public async Task OperatorEqualAndAdditionSubtraction_ReferenceType_Diagnostic()
+        {
+            // Doc states that if type behaves as a value-type and has addition/subtraction it might be safe.
+
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public class C
+{
+    public static bool operator [|==|](C left, C right) => true;
+    public static bool operator !=(C left, C right) => true;
+
+    public static C operator +(C left, C right) => left;
+    public static C operator -(C left, C right) => left;
```

Comment:
I kept the legacy implementation but we could have some FPs here so it might be even better to always ignore cases where there is an add/remove operator, wdyt?

Reply:
If the type is behaving as a number it should be IComparable(Of Itself) and IEquatable(Of Itself).  Since IEquatable will make the warning go away, I don't think there needs to be any special treatment for these operators.

Reply:
That's ok for me.